### PR TITLE
AHOYAPPS-566: Enable bandwidth profile

### PR DESCRIPTION
**Pull request by timrozum**
https://issues.corp.twilio.com/browse/AHOYAPPS-566
https://issues.corp.twilio.com/browse/AHOYAPPS-582
https://issues.corp.twilio.com/browse/AHOYAPPS-587

#### Changes

1. Enable bandwidth profile.
1. Set video track priority to high if the video track is displayed in the main video view except if the reason is dominant speaker. The server manages video track priority for dominant speaker automatically because changing it locally is too slow.
1. Set max tracks to 10.
1. Changed default video codec from `H.264` to `VP8 Simulcast`.
1. I had to create my own VideoTrack, LocalVideoTrack, and RemoteVideoTrack so that my feature logic and UI code does not have to be concerned with the kind of video track. It can simply set priority to whatever it needs to be regardless if it is local or remote. Obviously if it is local it will result in no operation. All of this is very similar to what I had to do for Participants previously so it should be familiar.

To verify it works I temporarily set `maxSubscriptionBitrate` to 100 and then watched video tracks that are not high priority switch off (freeze).

There is a separate ticket to change UI for switched off video tracks.

I created a feature branch for the base branch because it can't be merged to master until SDK is released.

I'm sort of letting the dust settle on my room screen refactor before diving into tests so that's why I did not add tests here. And we are kind of looking into some tools to help with UI testing.

### CODE REVIEW

File: VideoApp/VideoApp/Stores/MainParticipant/MainParticipantStore.swift [hunk 1]
```diff
@@ -17,7 +17,26 @@
...
+            notificationCenter.post(name: .mainParticipantStoreUpdate, object: self)
+        }
+    }
+    private(set) var videoTrack: VideoTrack? {
+        didSet {
+            guard oldValue !== videoTrack else { return }
+
+            oldValue?.priority = nil
+        }
+    }
```
Comment (timrozum) [inline VideoApp/VideoApp/Stores/MainParticipant/MainParticipantStore.swift:39]: This file got quite a lot simpler even though it's doing more work because:

1. Using `didSet` is really nice for comparing new and old values. This kind of logic was in a separate function and I had to keep track of my own new and old state.
1. The update handling was really overly optimized and is not just more code but it was getting a little hard to follow. The logic to calculate the main participant is O(n) and the UI updates are tiny and it is only updating one view so really we can just process any update the same way.

File: VideoApp/VideoApp/Video/Participants/LocalParticipant.swift [hunk 2]
```diff
@@ -19,11 +19,12 @@ import TwilioVideo
 
 class LocalParticipant: NSObject, Participant {
     let identity: String
-    var cameraTrack: VideoTrack? { localCameraTrack }
+    let isDominantSpeaker = false
+    let isRemote = false
+    var cameraTrack: VideoTrack? { cameraManager?.track }
```
Comment (timrozum) [inline VideoApp/VideoApp/Video/Participants/LocalParticipant.swift:24]: There is a little noise here because I moved some variables. Wanted to list constants before vars for example.

File: VideoApp/VideoApp/Video/Participants/LocalParticipant.swift [hunk 3]
```diff
@@ -61,11 +61,11 @@ class LocalParticipant: NSObject, Participant {
                 
                 self.cameraManager = cameraManager
                 cameraManager.delegate = self
-                participant?.publishVideoTrack(cameraManager.track)
+                participant?.publishVideoTrack(cameraManager.track.track)
```
Comment (timrozum) [inline VideoApp/VideoApp/Video/Participants/LocalParticipant.swift:64]: `track.track` is a little odd but seemed like least bad option.

File: VideoApp/VideoApp/Video/Room/ConnectOptionsFactory.swift [hunk 4]
```diff
@@ -35,7 +35,11 @@ import TwilioVideo
                 localVerbosity: .minimal,
                 remoteVerbosity: .minimal
             )
-            
+            builder.bandwidthProfileOptions = BandwidthProfileOptions(
+                videoOptions: VideoBandwidthProfileOptions { builder in
+                    builder.mode = .collaboration
```
Comment (timrozum) [inline VideoApp/VideoApp/Video/Room/ConnectOptionsFactory.swift:40]: I did not specify custom render dimensions. I'm a little confused about if I need to do that because documentation kind of implies it's not important. I have more bandwidth profile tickets including one to add settings so we could always optimize it later.

File: VideoApp/VideoApp/Video/Room/ConnectOptionsFactory.swift [hunk 5]
```diff
@@ -35,19 +35,30 @@ import TwilioVideo
...
+                    builder.maxSubscriptionBitrate = 6_000
+                    builder.maxTracks = 10
+                    builder.dominantSpeakerPriority = .high
+                    builder.trackSwitchOffMode = .detected
+                    builder.renderDimensions = VideoRenderDimensions()
+                    builder.renderDimensions?.high = VideoDimensions(width: 1_280, height: 720)
+                    builder.renderDimensions?.low = VideoDimensions(width: 352, height: 288)
```
[the remark this answers is not in the document]
Comment (timrozum) [reply VideoApp/VideoApp/Video/Room/ConnectOptionsFactory.swift:47]: @aaalaniz @piyushtank @paynerc I did the best I could to make app match docs but I have the remaining open items above.

[the remark this answers is not in the document]
Comment (aaalaniz) [reply VideoApp/VideoApp/Video/Room/ConnectOptionsFactory.swift:47]: Hey @timrozum

Thanks for getting a jump on this. Here is my feedback about the BWP default behavior. In general, I think we should follow the guidance provided for [mobile browsers](https://www.twilio.com/docs/video/tutorials/developing-high-quality-video-applications#mobile-browser-collaboration-recommended-settings) and be consistent with the mobile browser BWP settings in the react app.

1. Let's confirm with @timmydoza what the max subscription bitrate is on mobile browsers. 6000 is well under the recommended 2500000 so let's just pick the same number set in the react app. cc @ceaglest 
2. I think 10 is too large a Max Tracks on mobile. I think we should follow the mobile browser guideline and go with 5 for now.
3. Let's just not set the values for dominantSpeakerPriority, trackSwitchOffMode, and renderDimension. This will result in "Server Defaults" being used as recommended in the guide.

We should also follow the guidance and publish the camera video track with low priority. I'm not sure if that can be incorporated into this PR or needs to be tracked in a separate ticket. 

I'll let @paynerc and @piyushtank discuss the changes in the video app with respect to the encodingParameters and format dimensions.

File: VideoApp/VideoApp/Stores/MainParticipant/MainParticipantStore.swift [hunk 6]
```diff
@@ -26,59 +45,29 @@ class MainParticipantStore {
...
+        mainParticipant = room.localParticipant
+        videoTrack = mainParticipant.mainVideoTrack
         update()
+        notificationCenter.addObserver(self, selector: #selector(update), name: .roomUpdate, object: room)
+        notificationCenter.addObserver(self, selector: #selector(update), name: .participantUpdate, object: nil)
+        notificationCenter.addObserver(self, selector: #selector(update), name: .participantsStoreUpdate, object: participantsStore)
     }
     
-    private func update() {
+    @objc private func update() {
```
Comment (paynerc) [inline VideoApp/VideoApp/Stores/MainParticipant/MainParticipantStore.swift:56]: Does this still need to be `@objc`?

Comment (timrozum) [reply VideoApp/VideoApp/Stores/MainParticipant/MainParticipantStore.swift:56]: Yeah the notification interface requires it I think because `#selector` relies on some objc magic. 😞

File: VideoApp/VideoApp/Stores/MainParticipant/MainParticipantStore.swift [hunk 7]
```diff
@@ -17,7 +17,26 @@
 import Foundation
 
 class MainParticipantStore {
-    private(set) var mainParticipant: Participant
+    private(set) var mainParticipant: Participant {
+        didSet {
+            videoTrack = mainParticipant.mainVideoTrack
+
+            if mainParticipant.isPinned || videoTrack === mainParticipant.screenTrack || !mainParticipant.isDominantSpeaker {
+                videoTrack?.priority = .high
```
[the remark this answers is not in the document]
Comment (timrozum) [reply VideoApp/VideoApp/Stores/MainParticipant/MainParticipantStore.swift:25]: Any participant can be pinned including local. And if there are no remote participants the local participant is displayed in the main view. 

In order to prevent problems like you mention (and keep this code pretty simple) I had to create a VideoTrack protocol similar to my Participant protocol. And then I have a LocalVideoTrack concrete and RemoteVideoTrack concrete very similar to what I did for participants previously. LocalVideoTrack does nothing when `priority` is set so totally safe. And of course RemoteVideoTrack calls down into TwilioVideo.RemoteVideoTrack when `priority` is set.

[the remark this answers is not in the document]
Comment (paynerc) [reply VideoApp/VideoApp/Stores/MainParticipant/MainParticipantStore.swift:25]: Yeah, I saw that later on as I scrolled. 👍

File: VideoApp/VideoApp/Video/Room/ConnectOptionsFactory.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (aaalaniz) [reply VideoApp/VideoApp/Video/Room/ConnectOptionsFactory.swift:47]: Discussed live.

Max subscription bitrate: Use default (2400 Kbps)

[the remark this answers is not in the document]
Comment (timrozum) [reply VideoApp/VideoApp/Video/Room/ConnectOptionsFactory.swift]: I think I'm all set and pushed changes. They were very simple so I will plan to merge after CI passes. 

I created this ticket for the encoding parameters and format dimensions because it seems I need to discuss more and it's kind of a different area of code:

https://issues.corp.twilio.com/browse/AHOYAPPS-670
